### PR TITLE
Fix #215 & #217

### LINKED ISSUE #215: Trampoline calls fail when prologue calls routine reading next instruction in 32-bit
## Description

Greetings again, @stevemk14ebr! I've spent some time investigating the reason behind failure of standard library functions on Linux in 32-bit architecture. This scenario is mentioned as _Issue A_ in #214 and it is implemented in the following unit test that is currently failing: (See https://github.com/stevemk14ebr/PolyHook_2_0/actions/runs/18063107063/job/51402168957#step:8:451)

https://github.com/stevemk14ebr/PolyHook_2_0/blob/10529b9add5f07de3345350acba816c71b7bf6f2/UnitTests/linux/TestDetourx86.cpp#L219-L229

The cause of this seg fault has nothing to do with offset size, as I speculated in the TODO comment. The real cause of the failure is a tiny routine (let's call it `get_next_instr_addr` for now), whose purpose it is to return the next instruction address:
```asm
___lldb_unnamed_symbol4945:
	mov    eax, dword ptr [esp]     
	ret   
```

The problem arises from the fact that when `get_next_instr_addr` is called from the trampoline, it will return address of the next instruction in trampoline, not in the original function. And this corrupt value then leads to seg faults. Please allow me to walk you through an example with illustrations.

## Example 1: Normal function call (before hooks)

Image

Here we can see that when we call `printf` (1) the second instruction that gets executed is `call 0xf7e89c27` (2), which is the `get_next_instr_addr` that I've described earlier. At this point, the `esp` stores a pointer  with address `0xf7d5eb89`, which is the third instruction in the `printf` function. The function then proceeds to use this address accordingly. So far so good. Now let's take a look at what happens when we have hooked the function using an x86 detour.

## Example 2: Hooked function call

Image

We start by casting our trampoline variable to the function type and calling it as usual (1). As expected, we land at the trampoline which executes the first two prologue instructions (2, 3, 4) before jumping to the original `printf` (5). However, please pay attention at what happens when `get_next_instr_addr` gets called (3). This routine does its job as expected, and returns next instruction address, but it is an address in trampoline, not in the original `printf` function. When we jump to original function after trampoline (5), the eax doesn't store the address of the third instruction in `printf` like in normal case. Hence, this eventually leads to a segfault.

## Remedies

None that come to my mind right now. But I think that Polyhook should somehow handle such situations. This situation might feel like a very obscure edge case, but is it really "edge" when nearly every function from standard 32-bit Linux libraries features such code? The only reason why other hooks like `recv` succeed in linux is because there are `push reg` instructions in between the `endbr32` and `call get_next_instr_addr` instruction, thus keeping `call get_next_instr_addr` in place, thus not corrupting the return value.

I would really love to hear your thoughts and ideas on how we should deal with this issue. I will be thinking about this as well in the coming days.

### PR DESCRIPTION
**Pull request by acidicoala**
While I was on a roll, I decided to create a branch from my branch so that I could work on a fix for #217 and a few other QoL improvements. This branch includes everything from #216 plus additional commits. I did it this way, so that you could chose if how you wish to review those fixes - in isolation or together. If you chose to approve and merge this PR, I will close the #216. If you chose to approve and merge the #216, I will close this PR and create a new one with the fix for #217 specifically.

This PR implements a fix for #217 as it was described in the first post of the issue, by replacing the `call rel32` instruction with `push imm32`.

With those 2 major issues fixed, there were still some minor issues preventing all tests from passing, so they were fixed in this branch as well. Specifically:
- **Added tests for issues #215 and #217**: We can't rely on Linux binaries to test those fixes since those binaries may or may not contain problematic instructions. Instead, inline assembly was used to construct appropriate test cases.
- **Added diagnostics in test mode**: When testing for various edge cases, like fixes for #215, #217, instruction translations, etc., we need to be able to confirm that said fixes actually took place. Until now this was only possible to do manually by inspecting the output logs. But now when building tests (`POLYHOOK_BUILD_DLL OFF`), a new `PLH_DIAGNOSTICS` definition will be compiled. When this definition is enabled, `IHook` class will contain a new `uint32_t m_diagnostics` member variable, which holds flags for various diagnostics. All flags are defined in the `Diagnostic` enum class in `Enums.hpp` (which itself is defined only if `PLH_DIAGNOSTICS` is defined). When a special fix gets applied, a corresponding diagnostic flag gets set, and then in tests we can verify that using the `IHook::hasDiagnostic(Diagnostic d)` helper. For instance, this has allowed me to uncover that no translation was actually taking place in the `dlmopen` instruction, so I disabled that test for now. For now these diagnostics are used in tests for #215 and #217 where they verify that the expected fixes did in fact take place. I can see other flags being added in the future, making the tests more reliable. Of course in library (non-test) builds, none of the diagnostic functions, enums and variables will be available, so this has no performance impact outside of tests. This design was inspired by the [nlohmann/json](https://github.com/nlohmann/json) library, which features a [JSON_DIAGNOSTICS](https://json.nlohmann.me/api/macros/json_diagnostics/) flag that works in a similar manner.
- Finally, the CI workflow was optimized to remove redundant & duplicate options, so keep the code lean and easily readable.

With all these changes, we should expect all tests to pass from this point forward.

### CODE REVIEW

File: polyhook2/Detour/x86Detour.hpp [hunk 1]
```diff
@@ -26,6 +23,10 @@ class x86Detour : public Detour {
     Mode getArchType() const override;
 
 protected:
+    void fixSpecialCases(insts_t& prologue);
+    void fixCallToRoutineReturningSP(Instruction& callInst, const insts_t& routine);
+    void fixInlineCallToReadSP(Instruction& callInst);
```
[the remark this answers is not in the document]
Comment (stevemk14ebr) [reply polyhook2/Detour/x86Detour.hpp:28]: Go for it

**Comment by acidicoala** [comment]
@stevemk14ebr, I've implemented the suggestion related to error checking in fixup functions. I've also took liberty to eliminate build warnings that were bugging me this whole time. This PR is now ready for your review.

**Review by stevemk14ebr — APPROVED** [review]
